packages/modules/product/src/services/product-module-service.ts

```diff
@@ -1759,4 +1801,35 @@ export default class ProductModuleService
 
     return variantsWithOptions
   }
+
+  protected static validateUniqueOptionsCombinations(
+    data: ((
+      | ProductTypes.CreateProductVariantDTO
+      | ProductTypes.UpdateProductVariantDTO
+    ) & { options: { id: string }[]; product_id: string })[],
+    variants: ProductVariant[]
+  ) {
+    for (const variantData of data) {
+      const existingVariant = variants.find((v) => {
+        if (
+          variantData.product_id! !== v.product_id ||
+          !variantData.options?.length
+        ) {
+          return false
+        }
+
+        return (variantData.options as unknown as { id: string }[])!.every(
+          ({ id: optionValueId }) =>
+            !!v.options.find((vo) => vo.id === optionValueId)
+        )
```

Comment:
nit(non-blocking): Can we improve readability here, think we try to jam a lot of things into a one-liner here 😄